libyul/optimiser/UnusedFunctionsCommon.h

```diff
@@ -0,0 +1,172 @@
+		if (std::holds_alternative<Assignment>(e))
+		{
+			if (std::holds_alternative<FunctionCall>(*std::get<Assignment>(e).value))
+			{
+				FunctionCall c = std::get<FunctionCall>(*std::get<Assignment>(e).value);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+		}
+		else if (std::holds_alternative<ExpressionStatement>(e))
+			if (std::holds_alternative<FunctionCall>(std::get<ExpressionStatement>(e).expression))
+			{
+				FunctionCall c = std::get<FunctionCall>(std::get<ExpressionStatement>(e).expression);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+	}
+	return false;
+}
+
+FunctionDefinition createReplacement(
+	FunctionDefinition& _old,
+	std::optional<std::vector<bool>> const& _usedParameters,
+	std::optional<std::vector<bool>> const& _usedReturnVariables,
+	NameDispenser& _nameDispenser,
+	YulString const& _newName)
+{
+	auto generateName = [&](TypedName t)
+	{
+		return TypedName{
+			t.location,
+			_nameDispenser.newName(t.name),
+			t.type
+		};
+	};
+
+	langutil::SourceLocation loc = _old.location;
+	TypedNameList functionParameters;
+	TypedNameList returnVariables;
+	TypedNameList renamedParameters = util::applyMap(_old.parameters, generateName);
+	TypedNameList reducedRenamedParameters;
+	TypedNameList renamedReturnVariables = util::applyMap(_old.returnVariables, generateName);
+	TypedNameList reducedRenamedReturnVariables;
+
+	if (_usedParameters)
+	{
+		std::vector<bool> const& mask = _usedParameters.value();
+		functionParameters = applyBooleanMask(_old.parameters, mask);
+		reducedRenamedParameters = applyBooleanMask(renamedParameters, mask);
+	}
+	else
+	{
+		functionParameters = _old.parameters;
+		reducedRenamedParameters = renamedParameters;
+	}
+
+	if (_usedReturnVariables)
+	{
+		std::vector<bool> const& mask = _usedReturnVariables.value();
+		returnVariables = applyBooleanMask(_old.returnVariables, mask);
+		reducedRenamedReturnVariables = applyBooleanMask(renamedReturnVariables, mask);
+	}
+	else
+	{
+		returnVariables = _old.returnVariables;
+		reducedRenamedReturnVariables = renamedReturnVariables;
+	}
+
+	FunctionDefinition newFunction{
+		loc,
+		_old.name,
+		renamedParameters,
+		reducedRenamedParameters,
+		{loc, {}} // body
+	};
+
+	_old.name = _newName;
+	_old.parameters = std::move(functionParameters);
+	_old.returnVariables = std::move(returnVariables);
+
+	std::swap(newFunction.parameters, renamedParameters);
+	std::swap(newFunction.returnVariables, renamedReturnVariables);
+
+	FunctionCall call{loc, Identifier{loc, _old.name}, {}};
+	for (auto const& p: reducedRenamedParameters)
+		call.arguments.emplace_back(Identifier{loc, p.name});
+
+	// Replace the body of `f_1` by an assignment which calls `f`, i.e.,
+	// `return_parameters = f(reduced_parameters)`
+	if (!_old.returnVariables.empty())
+	{
+		Assignment assignment;
+		assignment.location = loc;
+
+		// The LHS of the assignment.
+		for (auto const& r: reducedRenamedReturnVariables)
```

Comment:
Is the following correct?
```suggestion
		for (auto const& r: _old.returnVariables)
```

Reply:
Ah no, it should be `newFunction.returnVariables` right?

Reply:
No, `old` should be correct, because `new` is the linking function and has the full set of return variables.